Devirtualize AutogradMeta.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28543 Test TensorTypeSet instead of autograd_meta_ for variable-ness.
* **#28526 Devirtualize AutogradMeta.**
* #28582 Merge libc10{,_cuda,_hip}.so libraries into libtorch.so
* #28554 [DO NOT MERGE] Test with companion builder patch
* #28287 Merge Tensor and Variable types.
* #28282 Unfriend Variable factory functions.
* #28581 Don't clobber pytorch image with libtorch build.

Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Devirtualize AutogradMeta.

Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

yf225 reviewed (approved):
Everything is much simpler now since we merged libc10 and libtorch :D

ezyang committed:
Update on "Devirtualize AutogradMeta."


Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize AutogradMeta."


Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize AutogradMeta."


Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize AutogradMeta."


Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize AutogradMeta."


Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize AutogradMeta."


Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Devirtualize AutogradMeta."


Now that torch/csrc/ and c10/ get linked into the same library, I can now
reference autograd code directly from TensorImpl.  That means I don't need
a virtual interface for AutogradMeta.  Delete it and replace it with direct
accesses from the C++ file (cannot put them inline, since AutogradMeta
is still an *incomplete* type at the time we are defining TensorImpl).

Note that this arrangement, where AutogradMeta is defined after Tensor,
is the ONLY arrangement that works, under the constraint that Tensor::grad()
returns a Tensor& reference.  To implement this interface, we must store
an honest-to-goodness copy of Tensor in AutogradMeta.

One subtlety when dealing with AutogradMeta as an incomplete type is that
the *destructor* must be defined out-of-line (since the destructor for
unique_ptr<AutogradMeta> requires a complete type!)  I was also surprised
to find that the move constructor for unique_ptr also requires complete types.

One thing I didn't touch, is that AutogradMeta::requires_grad() is still virtual.
This is because the implementation differs on its view subclass.  I'm not really
sure the class hierarchy here is buying us all that much, but that's for another
day.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
NOTE: This automatically-posted comment is a test. Please ignore.

 (experimental) 0/10 flaky

kostmo commented:
## CircleCI build failures summary
As of commit 5710d3e:
* 0/10 recognized as flaky
* 10/10 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=5710d3eaa0c882d3c8f3f00008274083dd8217c4).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).
